File: frontend/src/services/documentTypeFieldService.ts
import api from './api';

export interface DocumentTypeField {
  id?: number;
  documentType: string;
  fieldKey: string;
  fieldLabel: string;
  fieldType: 'text' | 'number' | 'date' | 'select' | 'multiselect';
  isRequired: boolean;
  isOcrMappable: boolean;
  ocrPattern?: string;
  defaultValue?: string;
  validationRules?: string;
  fieldOptions?: string; // JSON string for select options
  displayOrder: number;
  isActive: boolean;
  description?: string;
}

export interface FieldOption {
  value: string;
  label: string;
}

export const documentTypeFieldService = {
  // Get all fields for a document type
  getFieldsForDocumentType: async (documentType: string): Promise<DocumentTypeField[]> => {
    const response = await api.get(`/document-type-fields/${documentType}`);
    return response.data;
  },

  // Get all fields including inactive (admin only)
  getAllFieldsForDocumentType: async (documentType: string): Promise<DocumentTypeField[]> => {
    const response = await api.get(`/document-type-fields/${documentType}/all`);
    return response.data;
  },

  // Create a new field configuration
  createField: async (field: DocumentTypeField): Promise<DocumentTypeField> => {
    const response = await api.post('/document-type-fields', field);
    return response.data;
  },

  // Update an existing field configuration
  updateField: async (id: number, field: DocumentTypeField): Promise<DocumentTypeField> => {
    const response = await api.put(`/document-type-fields/${id}`, field);
    return response.data;
  },

  // Delete a field configuration
  deleteField: async (id: number): Promise<void> => {
    await api.delete(`/document-type-fields/${id}`);
  },

  // Deactivate a field configuration
  deactivateField: async (id: number): Promise<DocumentTypeField> => {
    const response = await api.patch(`/document-type-fields/${id}/deactivate`);
    return response.data;
  },

  // Map OCR data to fields
  mapOcrData: async (documentType: string, ocrText: string): Promise<Record<string, string>> => {
    const response = await api.post(`/document-type-fields/${documentType}/map-ocr`, { ocrText });
    return response.data;
  },

  // Validate field values
  validateFields: async (
    documentType: string,
    fieldValues: Record<string, string>
  ): Promise<{ valid: boolean; message: string }> => {
    const response = await api.post(`/document-type-fields/${documentType}/validate`, fieldValues);
    return response.data;
  },

  // Parse field options JSON
  parseFieldOptions: (fieldOptions?: string): FieldOption[] => {
    if (!fieldOptions) return [];
    try {
      return JSON.parse(fieldOptions);
    } catch {
      return [];
    }
  },
};

export default documentTypeFieldService;

File: backend/src/main/java/com/bpdb/dms/service/DocumentTypeFieldService.java
package com.bpdb.dms.service;

import com.bpdb.dms.entity.DocumentTypeField;
import com.bpdb.dms.repository.DocumentTypeFieldRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

@Service
@Transactional
public class DocumentTypeFieldService {

    private static final Logger logger = LoggerFactory.getLogger(DocumentTypeFieldService.class);

    private final DocumentTypeFieldRepository repository;

    public DocumentTypeFieldService(DocumentTypeFieldRepository repository) {
        this.repository = repository;
    }

    /**
     * Get all active fields for a document type
     */
    public List<DocumentTypeField> getFieldsForDocumentType(String documentType) {
        return repository.findByDocumentTypeAndIsActiveTrueOrderByDisplayOrderAsc(documentType);
    }

    /**
     * Get all fields for a document type (including inactive)
     */
    public List<DocumentTypeField> getAllFieldsForDocumentType(String documentType) {
        return repository.findByDocumentTypeOrderByDisplayOrderAsc(documentType);
    }

    /**
     * Get field by document type and field key
     */
    public Optional<DocumentTypeField> getField(String documentType, String fieldKey) {
        return repository.findByDocumentTypeAndFieldKey(documentType, fieldKey);
    }

    /**
     * Create or update a field configuration
     */
    public DocumentTypeField saveField(DocumentTypeField field) {
        if (field.getId() == null) {
            // Check if field already exists
            if (repository.existsByDocumentTypeAndFieldKey(field.getDocumentType(), field.getFieldKey())) {
                throw new IllegalArgumentException(
                    String.format("Field '%s' already exists for document type '%s'", 
                        field.getFieldKey(), field.getDocumentType()));
            }
        }
        return repository.save(field);
    }

    /**
     * Delete a field configuration
     */
    public void deleteField(Long id) {
        repository.deleteById(id);
    }

    /**
     * Deactivate a field (soft delete)
     */
    public DocumentTypeField deactivateField(Long id) {
        DocumentTypeField field = repository.findById(id)
            .orElseThrow(() -> new IllegalArgumentException("Field not found: " + id));
        field.setIsActive(false);
        return repository.save(field);
    }

    /**
     * Map OCR extracted data to document type fields
     */
    public Map<String, String> mapOcrDataToFields(String documentType, String ocrText) {
        List<DocumentTypeField> fields = repository
            .findByDocumentTypeAndIsActiveTrueOrderByDisplayOrderAsc(documentType);
        
        return fields.stream()
            .filter(DocumentTypeField::getIsOcrMappable)
            .filter(field -> field.getOcrPattern() != null && !field.getOcrPattern().isBlank())
            .collect(Collectors.toMap(
                DocumentTypeField::getFieldKey,
                field -> extractValueFromOcr(ocrText, field.getOcrPattern()),
                (existing, replacement) -> existing // Keep first match
            ));
    }

    /**
     * Extract value from OCR text using pattern
     */
    private String extractValueFromOcr(String ocrText, String pattern) {
        if (ocrText == null || ocrText.isBlank() || pattern == null || pattern.isBlank()) {
            return null;
        }
        try {
            Pattern regex = Pattern.compile(pattern, Pattern.CASE_INSENSITIVE);
            java.util.regex.Matcher matcher = regex.matcher(ocrText);
            if (matcher.find()) {
                // Try to get the last group (usually the value)
                int groupCount = matcher.groupCount();
                if (groupCount > 0) {
                    String value = matcher.group(groupCount);
                    return value != null ? value.trim() : null;
                }
                return matcher.group(0);
            }
        } catch (Exception e) {
            logger.warn("Failed to extract value using pattern '{}': {}", pattern, e.getMessage());
        }
        return null;
    }

    /**
     * Validate field values against field configurations
     */
    public void validateFieldValues(String documentType, Map<String, String> fieldValues) {
        List<DocumentTypeField> fields = getFieldsForDocumentType(documentType);
        
        for (DocumentTypeField field : fields) {
            if (field.getIsRequired() && 
                (fieldValues == null || !fieldValues.containsKey(field.getFieldKey()) || 
                 fieldValues.get(field.getFieldKey()) == null || 
                 fieldValues.get(field.getFieldKey()).isBlank())) {
                throw new IllegalArgumentException(
                    String.format("Required field '%s' (%s) is missing", 
                        field.getFieldLabel(), field.getFieldKey()));
            }
        }
    }
}

